form checkout

desafree opened the pull request:
Created form checkout (submission functionality not present yet). Used plain input element+styled component instead of Material UI for the inputs that interact with the react-hook-form library

desafree committed:
feat:create checkout form page

-create form
-integrate form style
-integrate validation

desafree committed:
refactor: improve clarity and reusability

-remove unnecessary prop type in radio component
-remove unnecessary styled component in radio component
-add align prop to typography component
-create a file for regex phone number

desafree committed:
refactor: add translation payment method options

desafree committed:
refactor: improve clarity

-remove unnecessarey {} in attribute that is string
-remove unnecessary typography component in input component

desafree committed:
refactor: improve clarity

-remove unnecessary typography in radiobutton component
-replace multiple box components with stack in formoptions component

rams23 commented on src/_shared/form/components/Input/Input.tsx:
```diff
@@ -15,11 +17,13 @@ const Input: FC<Props> = ({ name, type }) => {
-    <div>
+    <InputStyled>
       <label htmlFor={name}>{name}</label>
       <input {...field} type={type} />
-      {invalid && <p>{error?.message}</p>}
-    </div>
+      {invalid && (
+        <Typography variant="body1">{error?.message || ""}</Typography>
+      )}
+    </InputStyled>
```
since we have it, I will use Mui text input

rams23 commented on src/_shared/form/components/Radio/RadioButton.tsx:
```diff
@@ -0,0 +1,18 @@
+  field: ControllerRenderProps<FieldValues, string>;
+}
+
+export const RadioButton: FC<Props> = ({ item, field }) => {
+  return (
+    <RadioButtonStyled>
+      <input {...field} type="radio" id={item} value={item} />
+      <label htmlFor={item}>{item}</label>
+    </RadioButtonStyled>
+  );
```
same as above

rams23 commented on src/checkout/components/BillingDetails/BillingDetails.tsx:
```diff
@@ -0,0 +1,24 @@
+        </Typography>
+      </Grid>
+      <Grid item xs={6}>
+        <Input name={translate("checkout.billing.name")} />
+      </Grid>
+      <Grid item xs={6}>
+        <Input name={translate("checkout.billing.email")} />
+      </Grid>
+      <Grid item xs={6}>
+        <Input name={translate("checkout.billing.phone")} />
```
the name of the input can not language-dependent, it is the name of the prop that will be added to the form values object, we neet to split the name from the translated label

rams23 commented on src/checkout/views/Checkout/Checkout.tsx:
```diff
@@ -0,0 +1,41 @@
+export const Checkout = () => {
+  const translate = useTranslate();
+
+  const formContextValue = useForm<FormData>({
```
Maybe we can change the FormData name with something like CheckOutData or similar

rams23 commented on src/checkout/utils/schemaCheckoutForm.ts:
```diff
@@ -0,0 +1,34 @@
+    .string()
+    .required()
+    .matches(/^[0-9]+$/, "Must be only digits")
+    .min(5, "Must be exactly 5 digits")
+    .max(5, "Must be exactly 5 digits"),
```
what if we will support more languages? also errors should be translated

rams23 commented on src/checkout/components/PaymentsDetails/PaymentsDetails.tsx:
```diff
@@ -0,0 +1,30 @@
+        <Radio
+          name={translate("checkout.payment.payment-method.title")}
+          list={[
+            translate("checkout.payment.payment-method.e-money"),
+            translate("checkout.payment.payment-method.cash-on-delivery"),
```
also the value selected from a radio button should not be language-dependent, we need to add value and label here

desafree committed:
refactor: improve modularity and clearity

-replace input elements with textfield mui component
-replace radio element with radio mui component
-replace single name props for label and name to different props for each one
-improve naming for checkoutdata type
-add traduction for errors
-remove traduction in name props

rams23 commented on src/_shared/form/components/Input/Input.tsx:
```diff
@@ -0,0 +1,41 @@
+      <Controller
+        name={name}
+        render={({ field: { onChange, onBlur, value } }) => (
+          <TextField
+            onChange={onChange}
+            onBlur={onBlur}
+            value={value}
+            label={label}
+          />
+        )}
+      ></Controller>
+      {invalid && error && error.message && (
+        <Typography variant="body1">{translate(error.message)}</Typography>
```
I think that TextField already support error message and style

rams23 commented on src/_shared/form/components/Input/Input.tsx:
```diff
@@ -0,0 +1,41 @@
+
+  return (
+    <Stack direction="column">
+      <Controller
+        name={name}
+        render={({ field: { onChange, onBlur, value } }) => (
```
is Controller required?

desafree replied:
React-hook-form docs suggest usign it with external library components. I think you can avoid using it with the inputProps of TextField but I think it will be a little bit tricker

rams23 replied:
but we are using useController in the same component to get the fields, info: let's check it

desafree replied:
Ohh yeah, using them both is redundant! I've removed useControlled and used only the Controller component!

desafree committed:
refactor: use error message in textfield instead of custom one

desafree committed:
refactor: remove redundant usecontroller

desafree committed:
Merge branch 'main' into checkoutform
